graphql_compiler/query_pagination/parameter_generator.py

```diff
@@ -1,4 +1,132 @@
+    pagination_field = vertex_partition.pagination_field
+    if vertex_partition.number_of_splits < 2:
+        raise AssertionError('Invalid number of splits {}'.format(vertex_partition))
+
+    # Find the FilterInfos on the pagination field
+    graphql_query_string = print_ast(query_ast)
+    query_metadata = graphql_to_ir(
+        schema_info.schema,
+        graphql_query_string,
+        type_equivalence_hints=schema_info.type_equivalence_hints
+    ).query_metadata_table
+    filter_infos = query_metadata.get_filter_infos(Location(tuple(vertex_partition.query_path)))
+    filters_on_field = [
+        filter_info
+        for filter_info in filter_infos
+        if filter_info.fields == (pagination_field,)
+    ]
+
+    # Get the value interval currently imposed by existing filters
+    integer_interval = get_integer_interval_for_filters_on_field(
+        schema_info, filters_on_field, vertex_type.name, pagination_field, parameters)
+    field_value_interval = _convert_int_interval_to_field_value_interval(
+        schema_info, vertex_type.name, pagination_field, integer_interval)
+
+    # Compute parameters
+    if is_uuid4_type(schema_info, vertex_type.name, pagination_field):
+        uuid_int_universe = Interval(MIN_UUID_INT, MAX_UUID_INT)
+        integer_interval = intersect_int_intervals(integer_interval, uuid_int_universe)
+
+        int_value_splits = (
+            integer_interval.lower_bound + int(
+                float(measure_int_interval(integer_interval) * i / vertex_partition.number_of_splits))
+            for i in range(1, vertex_partition.number_of_splits)
+        )
+        return (
+            convert_int_to_field_value(schema_info, vertex_type.name, pagination_field, int_value)
+            for int_value in int_value_splits
+        )
+    else:
+        quantiles = schema_info.statistics.get_field_quantiles(
+            vertex_type.name, pagination_field)
+        if quantiles is None or len(quantiles) <= vertex_partition.number_of_splits:
+            raise AssertionError('Invalid vertex partition {}. Not enough quantile data.'
+                                 .format(vertex_partition))
+
+        # Since we can't be sure the minimum observed value is the
+        # actual minimum value, we treat values less than it as part
+        # of the first quantile. That's why we drop the minimum and
+        # maximum observed values from the quantile list.
+        proper_quantiles = quantiles[1:-1]
+
+        # Get the relevant quantiles (ones inside the field_value_interval)
+        min_quantile = 0
+        max_quantile = len(proper_quantiles)
+        if field_value_interval.lower_bound is not None:
+            min_quantile = bisect.bisect_left(proper_quantiles, field_value_interval.lower_bound)
+        if field_value_interval.upper_bound is not None:
+            max_quantile = bisect.bisect_left(proper_quantiles, field_value_interval.upper_bound)
```

Comment:
`bisect_left` is different from `bisect_right` only in how it treats edge cases, elements that are equal to an element in the list. I've arbitrarily chosen `bisect_left` and have been consistent with it. What this means is that a quantile value x separates the space into values less than **or equal** to x, and values greater than x. So what I need to do is:
- to make more precise the meaning of the quantiles
- to isolate this into a well-documented helper function explaining the choice of `bisect_left`

Reply:
I may be misunderstanding something. Please excuse my ignorance.

The docstring on `Interval` says that the interval endpoints are inclusive. My understanding, then, is that quantiles that may contain `field_value_interval.upper_bound` values should be preserved because the data we are looking for might be part of those quantiles. The inclusivity of both endpoints of `Interval` makes me think that we want `bisect_left` for the lower bound, and `bisect_right` for the upper bound, so that get the inclusive range of quantiles on both sides.

Let's hop on a quick call in the morning and figure out where my misunderstanding is?